NBitcoin.Altcoins/Stratis.cs

```diff
@@ -0,0 +1,660 @@
+					List<BlockHeaderWithTxCount> headersOff = new List<BlockHeaderWithTxCount>();
+					stream.ReadWrite(ref headersOff);
+					Headers.AddRange(headersOff.Select(h => h._Header));
+				}
+			}
+		}
+
+		public class StratisBlockSignature : IBitcoinSerializable
+		{
+			protected bool Equals(StratisBlockSignature other)
+			{
+				return Equals(signature, other.signature);
+			}
+
+			public override bool Equals(object obj)
+			{
+				if (ReferenceEquals(null, obj)) return false;
+				if (ReferenceEquals(this, obj)) return true;
+				if (obj.GetType() != this.GetType()) return false;
+				return Equals((StratisBlockSignature)obj);
+			}
+
+			public override int GetHashCode()
+			{
+				return (signature?.GetHashCode() ?? 0);
+			}
+
+			public StratisBlockSignature()
+			{
+				this.signature = new byte[0];
+			}
+
+			private byte[] signature;
+
+			public byte[] Signature
+			{
+				get
+				{
+					return signature;
+				}
+				set
+				{
+					signature = value;
+				}
+			}
+
+			internal void SetNull()
+			{
+				signature = new byte[0];
+			}
+
+			public bool IsEmpty()
+			{
+				return !this.signature.Any();
+			}
+
+			public static bool operator ==(StratisBlockSignature a, StratisBlockSignature b)
+			{
+				if (System.Object.ReferenceEquals(a, b))
+					return true;
+
+				if (((object)a == null) || ((object)b == null))
+					return false;
+
+				return a.signature.SequenceEqual(b.signature);
+			}
+
+			public static bool operator !=(StratisBlockSignature a, StratisBlockSignature b)
+			{
+				return !(a == b);
+			}
+
+			#region IBitcoinSerializable Members
+
+			public void ReadWrite(BitcoinStream stream)
+			{
+				stream.ReadWriteAsVarString(ref signature);
+			}
+
+			#endregion
+
+			public override string ToString()
+			{
+				return Encoders.Hex.EncodeData(this.signature);
+			}
+		}
+
+		/// <summary>
+		/// A POS block header, this will create a work hash based on the X13 hash algos.
+		/// </summary>
+		public class StratisBlockHeader : BlockHeader
+		{
+			const int CurrentVersion = 7;
+
+			private readonly X13 x13;
+
+			public StratisBlockHeader() : base()
+			{
+				this.x13 = new X13();
+			}
+
+			protected internal override void SetNull()
+			{
+				nVersion = CurrentVersion;
+				hashPrevBlock = 0;
+				hashMerkleRoot = 0;
+				nTime = 0;
+				nBits = 0;
+				nNonce = 0;
+			}
+
+			private byte[] CalculateHash(byte[] data, int offset, int count)
+			{
+				byte[] hashData = data.Skip(offset).Take(count).ToArray();
+				uint256 hash = null;
+				if (this.nVersion > 6)
+					hash = Hashes.Hash256(hashData);
+				else
+				{
+					var x13 = new X13();
+					hash = new uint256(this.x13.ComputeBytes(hashData));
+				}
+				return hash.ToBytes();
+			}
+
+			protected override HashStreamBase CreateHashStream()
+			{
+				return BufferedHashStream.CreateFrom(CalculateHash);
+			}
+
+			public override uint256 GetPoWHash()
+			{
+				
+				return new uint256(this.x13.ComputeBytes(this.ToBytes()));
```

Comment:
I am not familiar with Stratis so, this is a question: doesn't it depend on the `nVersion`?

Reply:
Hmmm, in looking at this I question that as well as CalculateHash checks the version. I double checked the stratis code (see below) and it matches what I have here so maybe we need @dangershony to weigh in on whether this is on purpose or not.
https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/src/NBitcoin/BlockStake.cs#L295-#L303

Reply:
My guess is Stratis POW is always using X13 so doesn't need to look at version but will get Dan to confirm.